catalyst-gateway-crates/c509-certificate/examples/cli/main.rs

```diff
@@ -0,0 +1,338 @@
+//! C509 certificate CLI
+
+use std::{
+    fs::{self, File},
+    io::Write,
+    path::PathBuf,
+};
+
+use asn1_rs::{oid, Oid};
+use c509_certificate::{
+    c509_big_uint::UnwrappedBigUint,
+    c509_extensions::Extensions,
+    c509_issuer_sig_algo::IssuerSignatureAlgorithm,
+    c509_name::{rdn::RelativeDistinguishedName, Name, NameValue},
+    c509_subject_pub_key_algo::SubjectPubKeyAlgorithm,
+    c509_time::Time,
+    signing::{PrivateKey, PublicKey},
+    tbs_cert::TbsCert,
+};
+use chrono::{DateTime, Utc};
+use clap::Parser;
+use hex::ToHex;
+use minicbor::Decode;
+use rand::Rng;
+use serde::{Deserialize, Serialize};
+
+/// Commands for C509 certificate generation, verification and decoding
+#[derive(Parser)]
+#[command(version, about, long_about = None)]
+enum Cli {
+    /// Generate C509 certificate, if private key is provided, self-signed certificate
+    /// will be generated.
+    Generate {
+        /// JSON file with information to create C509 certificate.
+        #[clap(short = 'f', long)]
+        json_file: PathBuf,
+        /// Optional output path that the generated C509 will be written to.
+        #[clap(short, long)]
+        output: Option<PathBuf>,
+        /// Optional private key file, if provided, self-signed certificate will be
+        /// generated. Currently support only PEM format.
+        #[clap(long)]
+        private_key: Option<PathBuf>,
+        #[clap(long)]
+        /// Optional key type.
+        key_type: Option<String>,
+    },
+
+    /// C509 certificate signature verification.
+    Verify {
+        /// C509 certificate file
+        #[clap(short, long)]
+        file: PathBuf,
+        /// Public key file. Currently support only PEM format.
+        #[clap(long)]
+        public_key: PathBuf,
+    },
+
+    /// Decode C509 certificate back to JSON.
+    Decode {
+        /// C509 certificate file.
+        #[clap(short, long)]
+        file: PathBuf,
+        /// Optional output path of C509 certificate information in JSON format.
+        #[clap(short, long)]
+        output: Option<PathBuf>,
+    },
+}
+
+impl Cli {
+    /// Function to execute the commands.
+    pub(crate) fn exec() -> anyhow::Result<()> {
+        let cli = Cli::parse();
+
+        match cli {
+            Cli::Generate {
+                json_file,
+                output,
+                private_key,
+                key_type,
+            } => {
+                let sk = match private_key {
+                    Some(key) => Some(PrivateKey::from_file(key)?),
+                    None => None,
+                };
```

Comment:
```suggestion
                let sk = private_key
                    .map(PrivateKey::from_file)
                    .transpose()?;
```

Reply:
I think this syntax is hard to interpret, will keep it this way, but thanks for the suggestion!